src/TreeDataGrid.tsx

```diff
@@ -0,0 +1,430 @@
+      if (isReadonlyArray(rows)) {
+        flattenedRows.push(...rows);
+        return;
+      }
+      Object.keys(rows).forEach((groupKey, posInSet, keys) => {
+        // TODO: should users have control over the generated key?
+        const id = parentId !== undefined ? `${parentId}__${groupKey}` : groupKey;
+        const isExpanded = expandedGroupIds.has(id);
+        const { childRows, childGroups, startRowIndex } = rows[groupKey];
+
+        const groupRow: GroupRow<R> = {
+          id,
+          parentId,
+          groupKey,
+          isExpanded,
+          childRows,
+          level,
+          posInSet,
+          startRowIndex,
+          setSize: keys.length
+        };
+        flattenedRows.push(groupRow);
+        allGroupRows.add(groupRow);
+
+        if (isExpanded) {
+          expandGroup(childGroups, id, level + 1);
+        }
+      });
+    };
+
+    expandGroup(groupedRows, undefined, 0);
+    return [flattenedRows, isGroupRow];
+
+    function isGroupRow(row: R | GroupRow<R>): row is GroupRow<R> {
+      return allGroupRows.has(row);
+    }
+  }, [expandedGroupIds, groupedRows, rawRows]);
+
+  const rowHeight = useMemo(() => {
+    if (typeof rawRowHeight === 'function') {
+      return ({ row, type }: RowHeightArgs<R | GroupRow<R>>): number => {
+        if (isGroupRow(row)) {
+          return rawRowHeight({ type: 'GROUP', row });
+        }
+        return rawRowHeight({ row, type });
+      };
+    }
+
+    return rawRowHeight;
+  }, [isGroupRow, rawRowHeight]);
+
+  const getParentRowAndIndex = useCallback(
+    (row: R | GroupRow<R>) => {
+      const rowIdx = rows.indexOf(row);
+      for (let i = rowIdx - 1; i >= 0; i--) {
+        const parentRow = rows[i];
+        if (isGroupRow(parentRow) && (!isGroupRow(row) || row.parentId === parentRow.id)) {
+          return [parentRow, i] as const;
+        }
+      }
+
+      return undefined;
+    },
+    [isGroupRow, rows]
+  );
+
+  const rowKeyGetter = useCallback(
+    (row: R | GroupRow<R>) => {
+      if (isGroupRow(row)) {
+        return row.id;
+      }
+
+      if (typeof rawRowKeyGetter === 'function') {
+        return rawRowKeyGetter(row);
+      }
+
+      const parentRowAndIndex = getParentRowAndIndex(row);
+      if (parentRowAndIndex !== undefined) {
+        const { startRowIndex, childRows } = parentRowAndIndex[0];
+        const groupIndex = childRows.indexOf(row);
+        return startRowIndex + groupIndex + 1;
+      }
+
+      return rows.indexOf(row);
+    },
+    [getParentRowAndIndex, isGroupRow, rawRowKeyGetter, rows]
+  );
+
+  const selectedRows = useMemo((): Maybe<ReadonlySet<Key>> => {
+    if (rawSelectedRows == null) return null;
+
+    assertIsValidKeyGetter<R, K>(rawRowKeyGetter);
+
+    const selectedRows = new Set<Key>(rawSelectedRows);
+    for (const row of rows) {
+      if (isGroupRow(row)) {
+        const isGroupRowSelected = row.childRows.every((cr) =>
+          rawSelectedRows.has(rawRowKeyGetter(cr))
+        );
```

Comment:
Does this handle more than one level of nested rows?

Reply:
It does as `row.childRows` contains all the raw rows within that group. So for example if we have 

```
groupLevel1
  - subGroupLevel1
    - row1
    - row2
  - subGroupLevel2
    - row3
```

Then 

```
groupLevel1,childRows == [row1, row2, row3]
subGroupLevel1,childRows == [row1, row2]
subGroupLevel2,childRows == [row3]
```